apps/aecore/src/aec_consensus_hc.erl

```diff
@@ -447,10 +452,92 @@ pc_start_height() ->
+    aeu_ets_cache:get(?ETS_CACHE_TABLE, child_epoch_length, Fun).
+
+child_block_time() ->
+    Fun = fun() -> get_consensus_config_key([<<"child_block_time">>]) end,
+    aeu_ets_cache:get(?ETS_CACHE_TABLE, child_block_time, Fun).
+
+is_new_epoch(Height) ->
+    Height rem child_epoch_length() == 0.
+
+hash_to_int(Hash) ->
+    {ok, Hash2IntFun} = aeu_ets_cache:lookup(?ETS_CACHE_TABLE, hash_to_int),
+    Hash2IntFun(Hash).
+
+
+entropy(Block, Height) ->
+    case aeu_ets_cache:lookup(?ETS_CACHE_TABLE, seed) of
+        {ok, {_OldState, Seed, Height, _OldPCHeight}} ->
+            Seed;
+        _Other ->
+            PCHeight = aec_parent_chain_block:height(Block),
+            Hash = aec_parent_chain_block:hash(Block),
+            case is_new_epoch(Height) of
+                true ->
+                    case set_new_entropy(Hash, Height, PCHeight) of
+                        {SeedState, _OldSeed, _OldHeight, SeedHeight} when PCHeight > SeedHeight ->
+                            %% The finality for the parent chain has not been reached
+                            next_entropy_from_seed(SeedState, Height, SeedHeight);
+                        _ ->
+                            Hash
+                    end;
+                _ ->
+                    next_entropy(Hash, Height, PCHeight)
+            end
+    end.
+
+set_new_entropy(Hash, Height, PCHeight) ->
+    case aec_parent_chain_cache:get_block_by_height(PCHeight + pc_finality()) of
+        {error, not_in_cache} ->
+            case aeu_ets_cache:lookup(?ETS_CACHE_TABLE, seed) of
+                error ->
+                    lager:error("Parent finality not reached for height ~p", [PCHeight]),
+                    aec_conductor:throw_error(parent_chain_block_not_synced);
+                {ok, {_OldState, _OldSeed, _OldHeight, _OldPCHeight} = Old} ->
+                    lager:warning("Parent finality not reached for parent height ~p", [PCHeight]),
+                    Old
+            end;
+        {ok, _Block} ->
+            lager:info("New epoch at parent height ~p", [PCHeight]),
+            Seed = hash_to_int(Hash),
+            State = rand:seed_s(exsss, Seed),
+            Result = {State, Hash, Height, PCHeight},
+            aeu_ets_cache:put(?ETS_CACHE_TABLE, seed, Result),
+            Result
+    end.
+
+next_entropy(Hash, Height, PCHeight) ->
+    {NewState, _Seed, _Height, NewPCHeight} = case aeu_ets_cache:lookup(?ETS_CACHE_TABLE, seed) of
+                                {ok, {_State, _OldSeed, _OldHeight, PCHeight} = Result} ->
+                                    Result;
+                                _ ->
+                                    lager:warning("Trying to set entropy for parent height ~p", [PCHeight]),
+                                    set_new_entropy(Hash, Height, PCHeight)
+                              end,
+    next_entropy_from_seed(NewState, Height, NewPCHeight).
+
+next_entropy_from_seed(State, Height, PCHeight) ->
```

Comment:
Connected to remarks of @Hans... would it not be better to prepare for taking the entropy from the first block of a new parent-chain epoch. Given that this PR is about adding epochs, we can use these epochs to pick the entropy from in the first block of such epochs.

(Clearly, if the code should still sync each block, then the epoch length is just 1)

Reply:
Currently it sets the entropy for the first block of the new child epoch by reading the parent hash from the cache.  The pc_height function decides which parent hash to choose. Is there something wrong with the logic?